### PR TITLE
Fix mypy errors

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
Fixed all the mypy errors

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
^^^

## Tracking Issue
Fixes https://github.com/flyteorg/flyte/issues/1030

## Follow-up issue
NA

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,11 +35,9 @@ fmt: ## Format code with black and isort
...
 .PHONY: lint
 lint: ## Run linters
-	mypy flytekit/core || true
-	mypy flytekit/types || true
-	mypy tests/flytekit/unit/core || true
-	# Exclude setup.py to fix error: Duplicate module named "setup"
-	mypy plugins --exclude setup.py || true
+	mypy flytekit/core
+	mypy flytekit/types
+	mypy --allow-empty-bodies --allow-redefinition tests/flytekit/unit/core
```
Comment (eapolinario) [inline Makefile:40]: why do we need those flags?

Comment (pingsutw) [reply Makefile]: In flytekit tests, many functions' body is empty, like [this](https://github.com/flyteorg/flytekit/blob/8c6764c299427cc5e4bda4406b2396082e099582/tests/flytekit/unit/core/test_docstring.py#L31), so we add `allow-empty-bodies` here.
I removed allow-redefinition, and updated the comment as well.

File: flytekit/core/data_persistence.py [hunk 2]
```diff
@@ -58,7 +58,7 @@ class DataPersistence(object):
     Base abstract type for all DataPersistence operations. This can be extended using the flytekitplugins architecture
     """
 
-    def __init__(self, name: str, default_prefix: typing.Optional[str] = None, **kwargs):
+    def __init__(self, name: str = "", default_prefix: typing.Optional[str] = None, **kwargs):
```
Comment (pingsutw) [inline flytekit/core/data_persistence.py:61]: here

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]

File: flytekit/core/base_task.py [hunk 3]
```diff
@@ -368,7 +374,7 @@ def __init__(
         self,
         task_type: str,
         name: str,
-        task_config: T,
+        task_config: Optional[T],
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/core/base_task.py:377]: Regular task doesn't have task_config. https://github.com/flyteorg/flytekit/blob/e4d7649b440e1985b3ecc1d2ecfe9bb45612e444/flytekit/core/task.py#L78

File: flytekit/core/base_task.py [hunk 4]
```diff
@@ -171,7 +177,7 @@ def __init__(
         FlyteEntities.entities.append(self)
 
     @property
-    def interface(self) -> Optional[_interface_models.TypedInterface]:
+    def interface(self) -> _interface_models.TypedInterface:
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/core/base_task.py:180]: Updated Constructor. the interface should not be None, although the input and output are none. https://github.com/flyteorg/flytekit/blob/8c7208282c8226099debe2f065a36590b35c89c6/flytekit/models/task.py#L449

File: flytekit/core/interface.py [hunk 5]
```diff
@@ -28,8 +28,8 @@ class Interface(object):
 
     def __init__(
         self,
-        inputs: typing.Optional[typing.Dict[str, Union[Type, Tuple[Type, Any]], None]] = None,
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/core/interface.py:31]: [Here](https://github.com/flyteorg/flytekit/blob/e4d7649b440e1985b3ecc1d2ecfe9bb45612e444/flytekit/core/interface.py#L304)

File: flytekit/core/testing.py [hunk 6]
```diff
@@ -9,7 +10,7 @@
 
 
 @contextmanager
-def task_mock(t: PythonTask) -> MagicMock:
+def task_mock(t: PythonTask) -> typing.Iterator[MagicMock]:
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/core/testing.py]: updated it. tyty

File: flytekit/core/interface.py [hunk 7]
```diff
@@ -28,8 +28,8 @@ class Interface(object):
...
-        outputs: typing.Optional[typing.Dict[str, Type]] = None,
+        inputs: Optional[Dict[str, Type]] | Optional[Dict[str, Tuple[Type, Any]]] = None,
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/core/interface.py]: it works. I changed to union to make code style consistency in the code base.

[the remark this answers is not in the document]
Comment (eapolinario) [reply flytekit/core/interface.py]: This is a python 3.10 only feature, no? Described in https://peps.python.org/pep-0604/

**Review by wild-endeavor — APPROVED** [review]

**Comment by wild-endeavor** [comment]
@eapolinario let's not include this in 1.2.x

**Pull request merged**